crates/bitwarden-ipc/src/ipc_client.rs

```diff
@@ -1,233 +1,288 @@
+    fn from(value: ReceiveError) -> Self {
+        match value {
+            ReceiveError::Channel(e) => TypedReceiveError::Channel(e),
+            ReceiveError::Timeout(e) => TypedReceiveError::Timeout(e),
+            ReceiveError::Cancelled => TypedReceiveError::Cancelled,
+        }
+    }
 }
 
 impl<Crypto, Com, Ses> IpcClient<Crypto, Com, Ses>
 where
     Crypto: CryptoProvider<Com, Ses>,
     Com: CommunicationBackend,
-    Ses: SessionRepository<Session = Crypto::Session>,
+    Ses: SessionRepository<Crypto::Session>,
 {
     pub fn new(crypto: Crypto, communication: Com, sessions: Ses) -> Arc<Self> {
         Arc::new(Self {
             crypto,
             communication,
             sessions,
+
+            incoming: RwLock::new(None),
+            cancellation_token: RwLock::new(None),
         })
     }
 
+    pub async fn start(self: &Arc<Self>) {
+        let cancellation_token = CancellationToken::new();
+        let mut self_cancellation_token = self.cancellation_token.write().await;
+
+        *self_cancellation_token = Some(cancellation_token.clone());
+
+        let com_receiver = self.communication.subscribe().await;
+        let (client_tx, client_rx) = tokio::sync::broadcast::channel(CHANNEL_BUFFER_CAPACITY);
+
+        let mut client_incoming = self.incoming.write().await;
+        *client_incoming = Some(client_rx);
+        let _ = client_incoming;
+
+        let client = self.clone();
+        let future = async move {
+            loop {
+                select! {
+                    _ = cancellation_token.cancelled() => {
+                        log::debug!("Cancellation signal received, stopping IPC client");
+                        break;
+                    }
+                    received = client.crypto.receive(&com_receiver, &client.communication, &client.sessions) => {
+                        match received {
+                            Ok(message) => {
+                                if client_tx.send(message).is_err() {
+                                    log::error!("Failed to save incoming message");
+                                    break;
+                                };
+                            }
+                            Err(e) => {
+                                log::error!("Error receiving message: {:?}", e);
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+            log::debug!("IPC client shutting down");
+            client.stop().await;
+        };
+
+        #[cfg(not(target_arch = "wasm32"))]
+        tokio::spawn(future);
+
+        #[cfg(target_arch = "wasm32")]
+        wasm_bindgen_futures::spawn_local(future);
+    }
+
+    pub async fn is_running(self: &Arc<Self>) -> bool {
+        let incoming = self.incoming.read().await;
+        let cancellation_token = self.cancellation_token.read().await;
+        incoming.is_some() && cancellation_token.is_some()
```

Comment:
I don't think this is an issue here, but we should be careful when holding two locks together, if any other parts of the code are doing the same thing in the opposite order, we can find ourselves with a deadlock. 

If we move the is_some to the variables, we avoid holding the locks and remove the problem from happening.
```suggestion
        let has_incoming = self.incoming.read().await.is_some();
        let has_cancellation_token = self.cancellation_token.read().await.is_some();
        has_incoming && has_cancellation_token
```

Reply:
Thanks for catching this, locks seem to be an easy footgun, it's not super obvious (unless you're familiar with how rust works) that this change would actually drop the locks. Fixed!